### ISSUE TITLE
Request threading issues?

### ISSUE THREAD

**Issue by chrismcv** [nats-io/nats.net.v1#55]
Hi Colin, 
I'm doing some fairly basic tests at the moment to explore throughput.  I've a system that I want to Request about 20 sources and aggregate the results.

If I just use publish (using the same approach as below), then I can comfortably publish 10,000 messages in <300ms on my PC.  So I put together the following snippet with Request, and I was expecting something similar, but am seeing `NATSTimeoutException` instead.

```
                Msg msg = null;
                ConnectionFactory cf = new ConnectionFactory();
                var sw = new Stopwatch();
                sw.Start();
                using (var c = cf.CreateConnection())
                {
                    var tasks = Enumerable.Range(0, 10000).Select(x =>
                         Task.Factory.StartNew(() =>
                         {
                           msg = c.Request("AppointmentSearch", new AppointmentSearch().ToMsg(), 10 * 1000);
                         })
                    );
                    Task.WaitAll(tasks.ToArray(), TimeSpan.FromSeconds(10));
                }
                return sw.ElapsedMilliseconds;
```

Looking at `gnatsd -V` I can see a throughput of about 1 message per second.  This doesn't seem right to me?  Is there anything I can look at?

**Comment by ColinSullivan1** [comment]
@chrismcv, Thank you for creating the issue - I'll take a look.  You should certainly be getting better performance than one msg/sec!  To sanity check, what happens when you make the requests serially, in a loop (versus creating tasks)?

**Comment by chrismcv** [comment]
I created the following method

```
 private Task<Msg> RequestAsync(IConnection c, string subject, byte[] data, int timeout)
        {
            var tcs = new TaskCompletionSource<Msg>();
            var ct = new CancellationTokenSource(timeout);
            ct.Token.Register(() => tcs.TrySetCanceled(), useSynchronizationContext: false);

            string inbox = c.NewInbox();

            var s = c.SubscribeAsync(inbox);
            s.AutoUnsubscribe(1);
            s.MessageHandler += (e, h) =>
            {
                tcs.TrySetResult(h.Message);
                try
                {
                    s.Unsubscribe();
                }
                catch { }
            };
            s.Start();
            c.Publish(subject, inbox, data);
            return tcs.Task;
        }
```

and using 

```
Msg msg = null;
                var res = new ConcurrentBag<AppointmentSearchResponse>();
                ConnectionFactory cf = new ConnectionFactory();
                var sw = new Stopwatch();
                sw.Start();
                using (var c = cf.CreateConnection())
                {
                    var tasks = Enumerable.Range(0, 10000).Select(async x =>
                    {
                        msg = await RequestAsync(c, "AppointmentSearch", new AppointmentSearch().ToMsg(), 10 * 1000);
                        res.Add(msg.As<AppointmentSearchResponse>());
                    });
                    Task.WaitAll(tasks.ToArray(), 100000);
                }
```

I get much better throughput - e.g. 600ms for 10,000.  I suspect the `SubscribeSync` in `request` is blocking other `Tasks` but `SubscribeAsync` isn't.

**Comment by ColinSullivan1** [comment]
Yes, SubscribeSync is the mechanism used to block the request, good catch!  I'll look into this further - Internally, SubscribeAsync with a monitor pulse could work better.

**Comment by ColinSullivan1** [comment]
@chrismcv I agree, but think this has more to do with the task scheduler itself, than NATS code.  I wrote a few tests using Tasks that failed quickly (running on VM), similar to your results.  The tasks weren't being scheduled in a performant manner, likely due to resources constraints.  However, when I moved down to using threads I found the performance I was expecting.  300 threads requesting, with the responder randomly delaying up to 5 seconds, and the tests passed about the same time as the delay, indicating there was little to no contention in request calls.  The responses were received were out of order, as expected, also indicating little contention.  Each subscription does have its own channel, so they shouldn't step on each other.  This is an interesting problem...  That being said, this needs to be addressed and I will continue looking.

**Comment by chrismcv** [comment]
@ColinSullivan1 did you try `ThreadPool.SetMinThreads(300, 300);` with tasks?  This seems to make it a bit more consistent for me.

**Comment by ColinSullivan1** [comment]
@chrismcv Thanks!  I did play around with that a bit, and it does help confirm my suspicions this is a resource problem.  I'll look into using raw threads under the hood to see if that'll make NATS more predictable under load...  To cover our bases, when you reduce the number of tasks in your test, do you get better throughput? (I'd assume so, but best to check).

**Comment by ColinSullivan1** [comment]
@chrismcv - I've added a few tests to illustrate the issue...

https://github.com/nats-io/csnats/blob/master/NATSUnitTests/UnitTestBasic.cs#L510
https://github.com/nats-io/csnats/blob/master/NATSUnitTests/UnitTestBasic.cs#L580

I'm finding that regardless of what is done internally in the NATS client (e.g. changing Tasks into Threads, blocking requests at a higher level using the `Monitor` API), I still run into the limitation and find Threads perform better than Tasks, as one would expect.  The good news is that with these tests, I'm very confident the `Connection.Request` API is sound with respect to thread safety. 

On the side, I found `TestRequestSafetyWithTasks` succeeded much more often by simply changing the subscriber's message delegate to create a thread rather than a task.  I'm thinking that at some point the task scheduler gets backed up delaying the NATS response.  

Regardless, your code is much more Task friendly using newer features (await).  We're still adhering to .NET 4.0 code compatibility.

Is this issue blocking you?   I'm wondering if we should chalk this up to scarcity of resources for the moment, but I can look deeper right now into this this is critical for you.

**Comment by liqdfire** [comment]
I would be more inclined to think that you are running into GC contention. Essentially by using Task.Run to create a Task for each item in your 10k range, you are allocating 10k instances of Tasks on the heap and the GC has to clean them up.

**Comment by ColinSullivan1** [comment]
This should be addressed with #106.
